### ISSUE TITLE
The reason why Oreo ROMs are not booting

### ISSUE THREAD

**Issue by vczilla** [chenxiaolong/DualBootPatcher#845]
So I did a write up on it on XDA.

I could copy and paste what I wrote over here or being the lazy, neer do well, good for nothing bastard that I am just have a good go at that prose:

https://forum.xda-developers.com/showpost.php?p=74373698&postcount=242

Pure poetry ain't it?
Anyways here's the clincher for y'all being even lazier than me:
It's due to changes that Google introduced to support early mounts of some partitions or more specifically that Oreo introduced a device tree backed,early mount thingy for a few select partitions, namely /system, /vendor and /odm.

Goddamn VIPs getting the special treatment.

So basically mbtool/init has to be modified to take into account that new early mount capability since it's hijacking mounting system,data and friends from the original fstab so that it can mount them from their location on the raw partitions which is a bind mount and not a device mount which I don't think is even possible from an device tree based fstab entry.

Anyways I'm rambling like an old coot but if you want that Oreo goodness now just have a read at my XDA post.

Or be patient and wait for our friend to patch DBP.

Back to my couch potato routine thing now.

**Comment by DimitriDokuchaev** [comment]
Hey @vczilla thank you for the post, I have been struggling and wondering why my Oreo roms were not working, i started thinking something changed the way they mount partitions in Oreo, I was not wrong, I just couldn't debug it and find out what it is, so I have to thank you, great work.

One question though, has this not been patched yet? Because I find it strange that yours is the only post talking about the issue.

I am going to read through your post see if i can patch my boot image accordingly, i may post here requesting help.

**Comment by vczilla** [comment]
I should've pointed out that all of this happens when using a data-slot.

I haven't tried any other install location.

To answer your question it's gotten worse on my OP5 actually.

Most Oreo roms moved fstab.qcom to /vendor/etc/
which completely prevents DBP from finding it. ( It just vibrates the phone 3 times and reboot).

Fortunately it's easy enough to fix:
a-DTB fix: remove the fstab entry from the device tree.

b-then add back fstab.qcom to the ramdisk ( where you can also change the verify mount flag for it). If your data partition is unencrypted don't forget to change the userdata mount flag to encryptable=footer.

c-in order for init to mount  fstab successfully you should also add back an init.qcom.rc to the ramdisk.

Here's its contents:
on fs
    wait /dev/block/platform/soc/1da4000.ufshc
    symlink /dev/block/platform/soc/1da4000.ufshc /dev/block/bootdevice
    mount_all fstab.qcom
    swapon_all fstab.qcom

Finally repack the boot image and replace the one DBP is using.

**Comment by vczilla** [comment]
Note that the formatting got messed up, I'll mark where each line starts for init.qcom.rc:
on fs
   wait...ufshc
   symlink.../dev/block/boot device
   mount_all fstab.qcom
   swap_all fstab.qcom

**Comment by DimitriDokuchaev** [comment]
Morning @vczilla 

Well this got complicated all of a sudden :/

So to clear a few things i am using data slot as well, I am not using a op5 but a op3t (shouldn't make a difference its an issue with dualboot patcher/how mount works on 8.0 not a device specific problem)

I have been messing with Magisk for two days and can't for the life of me do anything with it, apparently those commands do not work anymore (running magisk 15.3) i get stuck in the very beginning when extracting the boot.img since /data/magisk/magiskboot does not exist.

I am "like fine let me download the magisk15.3.zip and see if I can find magiskboot inside of it" and sure enough there it was. 

I copied both magiskinit and magiskboot and the boot.img i wanted work with to the internal storage of the phone, and went to terminal su ./magiskboot --extract boot.img and extract is not a recognized option, the syntax and commands must have changed since you have written the tutorial.

So i went to another route, used android image kitchen to extract the boot.img, which apparently worked, i got a ramdisk folder with android system files and i got a split_img folder which holds both the ramdisk.cpio.gz and zimage but i still don't have the dtb files.

I then stumbled across this https://github.com/PabloCastellano/extract-dtb to extract the DTBs and got myself two files extracted from zimage:
00_kernel
01_dtbdump_Qualcomm_Technologies,_Inc._MSM_8996_v3_+_PMI8996_MTP.dtb
02_dtbdump_Qualcomm_Technologies,_Inc._MSM_8996pro_+_PMI8996_MTP.dtb

now i am at a loss about what i do, GNURoot Debian refuses to install, and i am not entirely sure if these are the necessary dtb files.

Can you give me a hand with this, I plan on writing a proper tutorial once I figure how all this works.

Thank you in advance.

Regards
 
Dimitri

**Comment by DimitriDokuchaev** [comment]
Somehow some more progress was made, and I did different tests regarding all this, firstly i went to test the same roms with multirom.

They all crash due to the fact that they can't find fstab.qcom, so right now dualboot patcher and multirom suffer the exact same issue.

Although one thing caught my eye, a specific rom by the name o Dirty Unicorns, that rom actually booted in multirom (to crash right after :/ ) but something is different about that 8.1.0 rom.

So I unpacked the boot.img of said rom, and lo and behold, fstab.qcom and init.qcom.rc are both present on the ramdisk, I extracted the DTB from the boot.img, and they do not contain the "fstab" entry on that specific rom fstab.qcom and init.qcom.rc exist where they have always existed.

I went on and tried dualboot patcher on that rom, and even though those files are in the correct place and the dtb is clean, the rom does the same thing when booting as the other roms :/ 
vibrates 4 times and kicks you to recovery.

I am at a loss here, I am going to try to patch Pixel Experience with the fstab.qcom and init.qcom.rc from Dirty Unicorns and see the outcome, but right now I don't understand why DU which fulfills the requirements and it is supposed to be booting, does not.

**Comment by stevexyz** [comment]
Great news for now!!
Bravo!!
:)

**Comment by vczilla** [comment]
Well magiskboot --unpack boot.img will unpack the ramdisk, kernel and dtb from the boot.img.
Then use dtc to decompile the dtb and remove the fstab entry.

Then recompile it and replace the original dtb with it.

Then to add both our fstab.qcom and init.qcom.rc to the ramdisk ( after you made sure these files are not already present in the ramdisk ) first copy both files in the same directory where you unpacked the boot.img.

magiskboot --cpio ./ramdisk.cpio 'add 0740 fstab.qcom fstab.qcom' 'add 0740 init.qcom.rc init.qcom.rc'

Then repack everything:
magiskboot --repack boot.img
A new boot.img called new-boot.img will then be created.

**Comment by DimitriDokuchaev** [comment]
Hey @vczilla 

I have done all that obviously.

It still doesn't work, the difference is I have added the permissions differently for fstab.qcom and init.qcom.rc, I added 0640 not 0740, which ones should I use, because previously you had written 0640.

Can you please show me the contents of your fstab.qcom and init.qcom.rc, i think this is where my issue is.

Would be very grateful for your response, so we can close this once and for all.

Thank you

**Comment by vczilla** [comment]
fstab.qcom:
`# Android fstab file.
# The filesystem that contains the filesystem checker binary (typically /system) cannot
# specify MF_CHECK, and must come before any filesystems that do specify MF_CHECK

#TODO: Add 'check' as fs_mgr_flags with data partition.
# Currently we dont have e2fsck compiled. So fs check would failed.

# Non-A/B fstab.qcom variant
#<src>                                   <mnt_point>        <type> <mnt_flags and options>                          <fs_mgr_flags>
/dev/block/bootdevice/by-name/system     /system            ext4   ro,barrier=1,discard                             wait
/dev/block/bootdevice/by-name/userdata   /data              ext4   noatime,nosuid,nodev,barrier=1,noauto_da_alloc,discard wait,check,encryptable=footer
/dev/block/zram0                         none               swap   defaults                                         zramsize=536870912
/dev/block/bootdevice/by-name/misc       /misc              emmc   defaults                                         defaults
/dev/block/bootdevice/by-name/modem      /firmware          vfat   ro,shortname=lower,uid=1000,gid=1000,dmask=227,fmask=337,context=u:object_r:firmware_file:s0 wait
/dev/block/bootdevice/by-name/bluetooth  /bt_firmware       vfat   ro,shortname=lower,uid=1002,gid=3002,dmask=227,fmask=337,context=u:object_r:bt_firmware_file:s0 wait
/dev/block/bootdevice/by-name/dsp        /dsp               ext4   ro,nosuid,nodev,barrier=1                        wait
/dev/block/bootdevice/by-name/persist    /persist           ext4   noatime,nosuid,nodev,barrier=1                   wait,check
/dev/block/bootdevice/by-name/cache      /cache             ext4   noatime,nosuid,nodev,barrier=1                   wait,check
`

**Comment by vczilla** [comment]
**fstab.qcom:**

```
/dev/block/bootdevice/by-name/system     /system            ext4   ro,barrier=1,discard                             wait
/dev/block/bootdevice/by-name/userdata   /data              ext4   noatime,nosuid,nodev,barrier=1,noauto_da_alloc,discard wait,check,encryptable=footer
/dev/block/zram0                         none               swap   defaults                                         zramsize=536870912
/dev/block/bootdevice/by-name/misc       /misc              emmc   defaults                                         defaults
/dev/block/bootdevice/by-name/modem      /firmware          vfat   ro,shortname=lower,uid=1000,gid=1000,dmask=227,fmask=337,context=u:object_r:firmware_file:s0 wait
/dev/block/bootdevice/by-name/bluetooth  /bt_firmware       vfat   ro,shortname=lower,uid=1002,gid=3002,dmask=227,fmask=337,context=u:object_r:bt_firmware_file:s0 wait
/dev/block/bootdevice/by-name/dsp        /dsp               ext4   ro,nosuid,nodev,barrier=1                        wait
/dev/block/bootdevice/by-name/persist    /persist           ext4   noatime,nosuid,nodev,barrier=1                   wait,check
/dev/block/bootdevice/by-name/cache      /cache             ext4   noatime,nosuid,nodev,barrier=1                   wait,check
```

**Comment by weigeloveu** [comment]
Command should be
```
./magiskboot --cpio ramdisk.cpio fstab.qcom fstab.qcom
```
But it'll extract all entries, anything wrong with the command?

**Comment by weigeloveu** [comment]
```
./magiskboot --cpio ramdisk.cpio extract fstab.qcom fstab.qcom
```
Sorry forgot extract

**Comment by weigeloveu** [comment]
I've tested los 15.1,boot successful but camera fc.i don't know why.
And I noticed when i run this code 

```
dtc -O dts -I dtb -o dt.txt dtb
```
![图片发自简书App](http://upload-images.jianshu.io/upload_images/1498027-05b14f911a228e7e.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1080/q/50)

there are many warning if it's the reason?

**Comment by weigeloveu** [comment]
```
root@localhost:/sdcard# dtc -O dts -I dtb -o dt.txt dtb
Warning (unit_address_vs_reg): Node /soc has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,mdss_mdp@900000/qcom,smmu_mdp_unsec_cb has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,mdss_mdp@900000/qcom,smmu_rot_unsec_cb has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,mdss_mdp@900000/qcom,smmu_mdp_sec_cb has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,mdss_mdp@900000/qcom,smmu_rot_sec_cb has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/ad-hoc-bus has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,cpu-bwmon has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,cpu-clock-8996@ has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,mhi@0 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/qcom,mhi-rmnet@0 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/qcom,mhi-rmnet@1 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/qcom,mhi-uci@0 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/qcom,msm-thermal has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,msm-cpe-lsm@3 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/cpr3-ctrl@99e8000/thread@0 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/cpr3-ctrl@99e8000/thread@1 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/cpr3-ctrl@838000/thread@0 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /soc/apc0-pwrcl-mem-acc-regulator has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/apc1-perfcl-mem-acc-regulator has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/apcc-l3-mem-acc-regulator has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,kgsl-iommu has a reg or ranges property, but no unit name
Warning (unit_address_vs_reg): Node /soc/qcom,system-stats@68140 has a unit name, but no reg property
Warning (unit_address_vs_reg): Node /memory has a reg or ranges property, but no unit name
root@localhost:/sdcard#
```

**Comment by weigeloveu** [comment]
another question is i unpack the ramdisk of omni,but can't find the fstab and int.qcom.rc.

**Comment by DimitriDokuchaev** [comment]
hello @weigeloveu 

which rom are you trying to unpack a nougat rom?

Either way i haven't made much progress, but i have moved towards multirom instead of dualboot patcher, and unfortunately the issue is the same, lack of fstab in the ramdisk.

I have made a post about what i attempted here: https://github.com/Tasssadar/multirom/issues/123 no response.
made the same post over here as well, https://forum.xda-developers.com/showpost.php?p=75866088&postcount=2454 if you want to add to discussion please do.

But as far as i can tell this no longer works (adding the fstab to the ramdisk) i have tried with 2 roms one got me far into boot that i could see the bootsplash (pixel experience), the other (lineage os) broke my touchscreen completely (not even a factory flash fixed it, i had to do a few tricks to revive it).

On the multirom side of things, i was able to compile multirom twrp with 8.1 recovery, i was told this was enough for multirom to boot 8.1.0 roms without any modifications, it didn't work, multirom complained there was no fstab in the ramdisk.

With all this said Ill rest my case, hopefully developers of multirom and dualbootpatcher will fix this eventually.

Regards :)

**Comment by weigeloveu** [comment]
@DimitriDokuchaev 
Nougat rom needn't to unpack boot,just patch it with dbp.
I've tried los 15.1 works well except camera fc on secondary rom.
Multirom seems dead,i'm op3 also.
For talk about it in time,i made a tg group if you are interested in it.welcome to join in.
https://t.me/dualbootpatcher

**Comment by vasishath** [comment]
I managed to boot a GSI aosp image as a secondary rom on multirom
